### PR TITLE
Migrate onboarding, cfrs from feature flags to nimbus exp…

### PR DESCRIPTION
**Pull request by ionutbedregeanu**
For #6387

I will mark this pr as draft until onboarding experiments are created

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on `Show All Checks`,
2. click `Details` next to `build-focus-debug` or `build-klar-debug` for changes targeting Klar,
2. click `View task in Taskcluster`,
3. click the `Artifacts` row,
4. click to download any of the apks listed here which use an appropriate name for each CPU architecture.

### CODE REVIEW

**Comment by ionutbedregeanu** [comment]
@jeffreygee Can you help me with a logo build for [these specifications](https://developer.android.com/guide/topics/ui/splash-screen)?

I tried with that one used for the onboarding screen but it doesn't look good.

#### Nightly
https://user-images.githubusercontent.com/11731652/159221195-b3465cd9-d8c2-4a7d-98b0-1e8c65f3d4f5.mp4

#### Dev/Production
https://user-images.githubusercontent.com/11731652/159221210-2f639e9d-cf57-4cb9-93d9-70c5f2869817.mp4

#### Beta
https://user-images.githubusercontent.com/11731652/159221381-878f73a4-de7a-4d0d-9aff-5c07f9355cdf.mp4

**Comment by jeffreygee** [comment]
@ionutbedregeanu - is there an option to do a foreground and background layer?

Foreground:
Foreground Icon

Background (dark mode):
Background

--

Otherwise we can try using this one:
Focus_Splash

**Comment by ionutbedregeanu** [comment]
@jeffreygee If I am using the above foreground we don't have consistency between logos
Splash screen logo will different from that one from onboarding or home screen

https://user-images.githubusercontent.com/11731652/159673835-7714a1a7-8618-4277-8891-8d0b2dffb77f.mp4

https://user-images.githubusercontent.com/11731652/159661192-bb06dcb2-df16-4bba-b764-46fa0c11ad4b.mp4

I read the documentation again and it seems that the launcher icon can be used and this is the result


https://user-images.githubusercontent.com/11731652/159674768-b6d8b3c9-6bf8-473d-9583-41d34521ee0b.mp4

https://user-images.githubusercontent.com/11731652/159674317-bf069fae-100b-42e8-aaae-573896a312d0.mp4

https://user-images.githubusercontent.com/11731652/159676166-07950de7-e6c2-4848-9997-e41f6a270756.mp4

As you can see in Nightly we can't really distinguish the logo from the background which is not ok in my opinion and in beta the foreground icon has a white container
So I need your input to know which direction should I follow, the first one with the same logo for all builds or the second one?

**Comment by jeffreygee** [comment]
@ionutbedregeanu - i prefer the second version as well, but can we have different background colors per build?

**Comment by ionutbedregeanu** [comment]
> @ionutbedregeanu - i prefer the second version as well, but can we have different background colors per build?

yes

**Comment by jeffreygee** [comment]
> > @ionutbedregeanu - i prefer the second version as well, but can we have different background colors per build?
> 
> yes

are hex codes sufficient or do you need SVGs of the backgrounds?

If hex codes are sufficient:
Focus: as is
Beta & Nightly: #f0f0f4

**Comment by ionutbedregeanu** [comment]
UI test should pass now, so I will restart the CI

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.kt
Comment (Amejia481) [inline app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:89]: After some manual testing, I was experiencing an empty screen for a couple of seconds before, I seeing the onboarding UI. 
https://user-images.githubusercontent.com/773158/160718264-c810796f-66e3-4467-ae2e-de719d3d0418.mp4

**Review by Amejia481 — COMMENTED** [review]
Migrating to nimbus looks good to me, I'm bit worry about this behavior https://github.com/mozilla-mobile/focus-android/pull/6618/commits/33fa80c9c2676970b904021f1747bb9dc6b50086#r837965536

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.kt
Comment (ionutbedregeanu) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:89]: It seems that Google already knows about this, but marked it as expected behaviour https://issuetracker.google.com/issues/205021357

Comment (Amejia481) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:89]: I believe we may need to build our own splash screen, as many users will be affected by this issue. Do they have a fix on an alpha channel of the library?

**Comment by Amejia481** [comment]
@ionutbedregeanu, I think we are missing telemetry event, when we are not able to download experiments before the time limit https://github.com/mozilla-mobile/focus-android/issues/6387#issuecomment-1069596795, did we covered it in another PR? :) 

> we will need to record an telemetry event to indicate that we were unable to download.

**Comment by ionutbedregeanu** [comment]
> @ionutbedregeanu, I think we are missing telemetry event, when we are not able to download experiments before the time limit [#6387 (comment)](https://github.com/mozilla-mobile/focus-android/issues/6387#issuecomment-1069596795), did we covered it in another PR? :)
> 
> > we will need to record an telemetry event to indicate that we were unable to download.

Thanks! No, is not covered. I will address it in this pr

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.kt
Comment (ionutbedregeanu) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:89]: From what I read...they don't. Yeah, we can add our own splash screen and I will get documented about how can we disable the default one which is added starting with Android 12

Comment (Amejia481) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:89]: I think it will be the best call, we could move that work in separate PR to unblock the other changes from landing.

**Review by Amejia481 — APPROVED** [review]
This looks good in general, let's just move the splash commit to an separate PR to address https://github.com/mozilla-mobile/focus-android/pull/6618#discussion_r837965536

**Comment by ionutbedregeanu** [comment]
> @ionutbedregeanu, I think we are missing telemetry event, when we are not able to download experiments before the time limit [#6387 (comment)](https://github.com/mozilla-mobile/focus-android/issues/6387#issuecomment-1069596795), did we covered it in another PR? :)
> 
> > we will need to record an telemetry event to indicate that we were unable to download.

@jhugman Can you help me with the specifications(name, type) of this event?

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.kt
Comment (ionutbedregeanu) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:89]: From what I found until now we can't disable the default splash screen for android 12. To move that work in a separate pr means to also remove `setSplashScreenPreDrawListener()` from this pr, which will affect the entire pr.
In my opinion we should also land this commit and search after that for a solution, because it seems to happens when the app is started from other applications or android studio

File: nimbus.fml.yaml
Comment (jhugman) [inline nimbus.fml.yaml:10]: Nit: use snake case here.

Also, naming: I like that it's called `tabs`; but is it really `is-multi-tab` or is it just `is-enabled` (I'm confused by a tabs feature without multi-tabs).

File: nimbus.fml.yaml
Comment (jhugman) [inline nimbus.fml.yaml:30]: `types` is redundant.

```suggestion
objects: {}
enums: {}
```

File: app/src/main/java/org/mozilla/focus/utils/Features.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/utils/Features.kt:27]: Do all these flags need to go in to the `nimbus.fml.yaml` file?

File: app/src/main/java/org/mozilla/focus/navigation/MainActivityNavigation.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/navigation/MainActivityNavigation.kt:141]: This would indicate we need an `onboarding` feature with a `variable` of `is-enabled`.

File: app/src/main/java/org/mozilla/focus/fragment/UrlInputFragment.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/fragment/UrlInputFragment.kt:357]: The feature here: `onboardingWithCfrs` doesn't fit with what it replaced: `ARE_HOME_PAGE_PRO_TIPS_ENABLED`.

Perhaps this needs to be in a feature grouped together with the FML equivalent of `SEARCH_TERMS_OR_URL`.

File: app/src/main/java/org/mozilla/focus/experiments/NimbusSetup.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/experiments/NimbusSetup.kt:88]: The startup sequence here still needs some work.

For first run only, when the splash screen is showing: the `NimbusInterface.Observer` should:

1. listen for the end of the fetch.
2. **Important** if and only if the `REQUEST_TIME_OUT` has not been reached, then call `applyPendingExperiments()`
3. Once `onExperimentsApplied()` is called we should arrange for the splash screen to be dismissed. 

**Important**: For subsequent runs we should not be registering an observer at all, and just calling `applyPendingExperiments()` and then `fetchExperiments()`.

Given that so much of this is tied up with the splash screen, I'd put the default `applyPendingExperiments()`/fetchExperiments()` calls here, then do the work with the observer closer to the splash screen.

File: app/build.gradle
Comment (jhugman) [inline app/build.gradle:190]: The Experimenter team has requested we use `.yaml`.

```suggestion
    experimenterManifest = ".experimenter.yaml"
```

File: nimbus.fml.yaml
Comment (jhugman) [reply nimbus.fml.yaml:27]: I think I'd like @amedyne and/or product to review this file. This file contains the switches and dials that experiment owners are going to toggle. Their arrangement, default values and grouping **must** be understandable to them.

**Comment by jhugman** [comment]
> > @ionutbedregeanu, I think we are missing telemetry event, when we are not able to download experiments before the time limit [#6387 (comment)](https://github.com/mozilla-mobile/focus-android/issues/6387#issuecomment-1069596795), did we covered it in another PR? :)
> > > we will need to record an telemetry event to indicate that we were unable to download.
> 
> @jhugman Can you help me with the specifications(name, type) of this event?

I got lost in reviewing this PR, and almost forgot to answer the question.

This is a new Glean event, so you'll need to add to `metrics.yaml` and go through data-review etc.

I _think_ you'll likely want [a `timespan`][1] called `nimbus_initial_fetch`, with extras:

 * A boolean `success`  for whether or not the fetch finished before the timeout.

This will be data sensitivity of `technical`.

[1]: https://mozilla.github.io/glean/book/reference/metrics/timespan.html

File: nimbus.fml.yaml
Comment (amedyne) [reply nimbus.fml.yaml:27]: The experiment design that was discussed for this was for the following:
- control: (current experience, no new onboarding or cfrs)
- new onboarding only 
- new onboarding + CFRs
I believe that what @jhugman recommended [here](https://github.com/mozilla-mobile/focus-android/pull/6618#discussion_r842635499) simplifies the variables.

@athomasmoz or @ItsDevyani any additional thoughts?

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.kt
Comment (ionutbedregeanu) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:62]: @jeffreygee Do we want to display the splash screen only on the first run, at fresh install?

File: nimbus.fml.yaml
Comment (ionutbedregeanu) [reply nimbus.fml.yaml:10]: This feature is already in production, in this pr it was just migrated to FML. In my opinion we should create another ticket to address this comment and not add regressions

File: app/src/main/java/org/mozilla/focus/utils/Features.kt
Comment (ionutbedregeanu) [reply app/src/main/java/org/mozilla/focus/utils/Features.kt:27]: All of them can be covered by `onboarding` feature from` nimbus.fml.yaml` file

File: app/src/main/java/org/mozilla/focus/fragment/UrlInputFragment.kt
Comment (ionutbedregeanu) [reply app/src/main/java/org/mozilla/focus/fragment/UrlInputFragment.kt:357]: In our case cfrs were designed to replace home screen tips, so if the experiment is not enabled we should follow the old path

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.kt
Comment (ionutbedregeanu) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:62]: Starting with Android 12 a default splash screen is added and maybe we should keep the consistency across all devices

File: nimbus.fml.yaml
Comment (ionutbedregeanu) [reply nimbus.fml.yaml]: @jhugman At your suggestion I made a single feature

**Comment by ionutbedregeanu** [comment]
> > > @ionutbedregeanu, I think we are missing telemetry event, when we are not able to download experiments before the time limit [#6387 (comment)](https://github.com/mozilla-mobile/focus-android/issues/6387#issuecomment-1069596795), did we covered it in another PR? :)
> > > > we will need to record an telemetry event to indicate that we were unable to download.
> > 
> > 
> > @jhugman Can you help me with the specifications(name, type) of this event?
> 
> I got lost in reviewing this PR, and almost forgot to answer the question.
> 
> This is a new Glean event, so you'll need to add to `metrics.yaml` and go through data-review etc.
> 
> I _think_ you'll likely want [a `timespan`](https://mozilla.github.io/glean/book/reference/metrics/timespan.html) called `nimbus_initial_fetch`, with extras:
> 
> * A boolean `success`  for whether or not the fetch finished before the timeout.
> 
> This will be data sensitivity of `technical`.

@amedyne I think I should add this telemetry in separate ticket. What do you think?

**Comment by amedyne** [comment]
> Submitted https://github.com/mozilla-mobile/focus-android/issues/6847 for the additional telemetry.

**Comment by ionutbedregeanu** [comment]
@jhugman Can you please take a second look on this pr?

File: .experimenter.yaml
Comment (jhugman) [inline .experimenter.yaml:35]: This says it's YAML, yet it's JSON. This is unlikely your fault, but something to look into.

File: app/build.gradle
Comment (jhugman) [reply app/build.gradle:191]: Ah, so this needs to be YAML, as the suggested changes suggests.

File: app/src/main/java/org/mozilla/focus/Components.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/Components.kt:193]: Nice.

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.kt
Comment (jhugman) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:65]: > maybe we should keep the consistency across all devices

Then this is a product decision. Please defer to your product manager.

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:139]: The logic here is pretty convoluted. Quite a lot of state being tracked, but without the benefit. 

Given you don't seem to be dismissing the splashscreen early, this could be more easily implemented as 

1. show the splash screen
2. `fetchExperiments()`
3. wait `REQUEST_TIME_OUT`
4. `applyPendingExperiments()`
5. dismiss the splash screen.

For step 4, we can call `applyPendingExperiments()` even if the fetch hasn't finished.

This way we don't need to do anything with a `Nimbus.Observer`, much less, keep track of whether or not we've called `applyPendingExperiments()`.

File: nimbus.fml.yaml
Comment (jhugman) [reply nimbus.fml.yaml:10]: > create another ticket to address this comment and not add regressions
Agreed. Please file a migration ticket which should be done after once experiments and rollouts on `is_multi_tab` are complete.

File: app/src/main/java/org/mozilla/focus/contextmenu/ContextMenuCandidates.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/contextmenu/ContextMenuCandidates.kt:26]: Nice, using a local variable is preferred to calling `FocusNimbus.features.tabs.value().isMultiTab` all the time.

My preferred style would be get all the configuration out of Nimbus with one call to `value()`.

```kt
val config = FocusNimbus.features.onboarding.value()

if (config.isEnabled) {}

if (config.isCfrEnabled) {} 
```

A bit redundant for just one flag here though :)

File: app/src/main/java/org/mozilla/focus/cfr/CfrMiddleware.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/cfr/CfrMiddleware.kt:24]: lovely.

**Review by jhugman — COMMENTED** [review]
All done, bar the nits. Well done.

The one other extra this needs are calls to `recordExposure`.

When the user is exposed to a feature (e.g. sees the onboarding flow) you need to call

```kt
FocusNimbus.features.onboarding.recordExposure()
```

For context: most users who are enrolled into the onboarding experiments will never see the new onboarding flow, so we need to be able eliminate them from the experimental cohort.

Nimbus will look up if the user is enrolled in an experiment and record an exposure event with the branch that they're in. Nimbus is smart enough not to emit an exposure if the user isn't in an experiment, so the app doesn't have to keep track of experiments, just whether the user has seen the feature.

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:132]:
```suggestion
        FocusNimbus.features.tabs.recordExposure()
```

You don't need to detect if the user is in an experiment; Nimbus already knows that.

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:153]: When is `onPreDraw()` called? I'm not sure I understand what is happening here.

File: app/src/main/java/org/mozilla/focus/contextmenu/ContextMenuCandidates.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/contextmenu/ContextMenuCandidates.kt:26]: Nit: the `value()` is a config, not an experiment. Suggest renaming to `tabConfig` or `multiTabConfg`.

File: app/src/main/java/org/mozilla/focus/navigation/MainActivityNavigation.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/navigation/MainActivityNavigation.kt:141]: This is interesting. Given that `isEnabled` is part of the feature values, I think we'd want to record exposure whatever the value of `isEnabled`.

Additionally, I think the `FeatureHolder` assumes that `value()` is called _before_ `recordExposure()`, so you could do this:

```suggestion
        val isOnboarding = FocusNimbus.features.onboarding.value().isEnabled
        FocusNimbus.features.onboarding.recordExposure()
        val onboardingFragment = if (isOnboarding) {
```

or:

```kotlin
        val onboardingFragment = if (FocusNimbus.features.onboarding.value().isEnabled) {
            FocusNimbus.features.onboarding.recordExposure()
            OnboardingFragment.create()
        } else {
            FocusNimbus.features.onboarding.recordExposure()
            FirstrunFragment.create()
        }   
```

**Review by jhugman — COMMENTED** [review]
I'm still not sure that the `onPreDraw()`/startup logic is right. 

Happy to pair if need be.

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.kt
Comment (ionutbedregeanu) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:143]: https://developer.android.com/guide/topics/ui/splash-screen#:~:text=Starting%20in%20Android%2012%2C%20the,transition%20to%20your%20app%20itself.
_Keep the splash screen on-screen for longer periods
The splash screen is dismissed as soon as your app draws its first frame. If you need to load a small amount of data such as in-app settings from a local disk asynchronously, you can use [ViewTreeObserver.OnPreDrawListener](https://developer.android.com/reference/android/view/ViewTreeObserver.OnPreDrawListener) to suspend the app to draw its first frame._

Comment (ionutbedregeanu) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:143]: I applied your suggestion https://github.com/mozilla-mobile/focus-android/pull/6618#discussion_r852872902
After 2 seconds applyPendingExperiments() is called(no matter what) and the splash screen is closed

File: nimbus.fml.yaml
Comment (ionutbedregeanu) [reply nimbus.fml.yaml:10]: https://github.com/mozilla-mobile/focus-android/issues/6901

File: app/src/main/java/org/mozilla/focus/tabs/MergeTabsMiddleware.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/tabs/MergeTabsMiddleware.kt:39]: 💪

File: app/src/main/java/org/mozilla/focus/navigation/MainActivityNavigation.kt
Comment (jhugman) [inline app/src/main/java/org/mozilla/focus/navigation/MainActivityNavigation.kt:147]: 💪

File: app/src/main/java/org/mozilla/focus/contextmenu/ContextMenuCandidates.kt
Comment (jhugman) [reply app/src/main/java/org/mozilla/focus/contextmenu/ContextMenuCandidates.kt:29]: Recording the exposure is likely not the right thing to do here, since we're not exposing the menu candidates to the user.

**Review by jhugman — APPROVED** [review]
We're down to the smallest of problems/nits now; I'm not going to hold up on another review cycle!

Well done, and thank you for your patience in working with the Nimbus team on this.

**Comment by ionutbedregeanu** [comment]
> We're down to the smallest of problems/nits now; I'm not going to hold up on another review cycle!
> 
> Well done, and thank you for your patience in working with the Nimbus team on this.

Thank you for guiding me through this kind of tickets!

**Pull request merged**
